### DISCUSSION TITLE
Compiling tasmota32Cx fails

### DISCUSSION THREAD

**Discussion by helmar74** [arendst/Tasmota#21372] [Q&A] [answered]
### PROBLEM DESCRIPTION
Hi,
I want to compile the current development version (13.4.1.2) for a tasmota32 device.
I changed user_config_override (see attached file inside zip).
I'm able to compile esp8266 firmware, but esp32 fails with:

```
*** [.pio\build\tasmota32\src\tasmota.ino.cpp.o] Error 1
```
Has anyone an idea, which entry in user_config_override.h causes the issue for esp32?

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): ESP32
- [X] Tasmota binary firmware version number used: 13.4.1.2
  - [ ] Pre-compiled
  - [X] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**
[user_config_override.zip](https://github.com/arendst/Tasmota/files/15219905/user_config_override.zip)

**Comment by Jason2866** [comment]
Your Platformio setup is broken. Close VSC. Delete hidden folder `.platformio` Do a fresh clone of Tasmota
Start VSC open "folder" of the new fresh clone. Let VSC install Platformio. If everything is up again copy your `user_config_override.h`

**Comment by Jason2866** [comment]
Using a specific display driver does not work anymore. All displays are working via uDisplay and a display.ini.

**Comment by helmar74** [comment]
OK, is there a description available, how I can use uDisplay for my RA8876 Display?
What do I have to change in my user_config_ovverride to use RA8876 Display, SD CARD, JPEG Pictures and so on?

**Reply by sfromis** [reply]
That's not via settings in `user_config_override.h`. You need a [display descriptor](https://tasmota.github.io/docs/Universal-Display-Driver/#step-5-configure-the-displayini-descriptor-file) for your specific display, and it does [not appear](https://github.com/arendst/Tasmota/tree/development/tasmota/displaydesc) that someone has created that. Yet.

Configuring SD card is in standard template settings.

**Comment by helmar74** [comment]
It also looks like this define is causing an compile error

```
#define JPEG_PICTS 
```

Compile error:

```
D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_support/support_jpeg.ino: In function 'bool jpg2rgb888(const uint8_t*, size_t, uint8_t*, jpg_scale_t)':
D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_support/support_jpeg.ino:135:39: error: invalid conversion from 'uint32_t (*)(void*, size_t, uint8_t*, size_t)' {aka 'long unsigned int (*)(void*, unsigned int, unsigned char*, unsigned int)'} to 'jpg_reader_cb' {aka 'unsigned int (*)(void*, unsigned int, unsigned char*, unsigned int)'} [-fpermissive]
  135 |     if(esp_jpg_decode(src_len, scale, _jpg_read, _rgb_write, (void*)&jpeg) != ESP_OK){
      |                                       ^~~~~~~~~
      |                                       |
      |                                       uint32_t (*)(void*, size_t, uint8_t*, size_t) {aka long unsigned int (*)(void*, unsigned int, unsigned char*, unsigned int)}
In file included from C:/Users/helmar/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32/include/esp32-camera/conversions/include/img_converters.h:25,
                 from D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_support/support_jpeg.ino:24:
C:/Users/helmar/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32/include/esp32-camera/conversions/include/esp_jpg_decode.h:37:71: note:   initializing argument 3 of 'esp_err_t esp_jpg_decode(size_t, jpg_scale_t, jpg_reader_cb, jpg_writer_cb, void*)'
   37 | esp_err_t esp_jpg_decode(size_t len, jpg_scale_t scale, jpg_reader_cb reader, jpg_writer_cb writer, void * arg);
      |                                                         ~~~~~~~~~~~~~~^~~~~~
D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_xdrv_driver/xdrv_13_display.ino: In function 'void Draw_JPG_from_URL(char*, uint16_t, uint16_t, uint8_t)':
D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_xdrv_driver/xdrv_13_display.ino:2507:5: error: 'Draw_jpeg' was not declared in this scope
 2507 |     Draw_jpeg(mem, jpgsize, xp, yp, scale);
      |     ^~~~~~~~~
*** [.pio\build\tasmota32\src\tasmota.ino.cpp.o] Error 1
Compiling .pio\build\tasmota32\lib2ae\WiFiHelper\WiFiHelper_ESP32.cpp.o
=========================================================================================== [FAILED] Took 75.21 seconds ===========================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32      FAILED    00:01:15.209
====================================================================================== 1 failed, 0 succeeded in 00:01:15.209 ====================================================================================== 

```

This should be the codesection in xdrv_13_display.ino  (line 2507):

```
if (jpgsize) {
    Draw_jpeg(mem, jpgsize, xp, yp, scale);
  }
  if (mem) free(mem);
```


Is this also related to Display Driver?

**Reply by sfromis** [reply]
Did you make sure to have this?
`#define USE_UFILESYS`

**Reply by helmar74** [reply]
Yes it is in user_config_override:
![image](https://github.com/arendst/Tasmota/assets/25499089/5e887b95-27e1-4e5f-b58d-ec2d29700f35)

The firmware is not compiling if #define JPEG_PICTS  is enabled
If i comment it out, the compiling works...

**Comment by gemu2015** [comment]
for core3 you have to change this line  in support_jpeg.ino (change uint32_t to size_t) 
static size_t _jpg_read(void * arg, size_t index, uint8_t *buf, size_t len)
it then it will compile. 
however jpeg decoder is broken with core3 and will crash.
the jpeg decoder used to be in flash and now is in ROM which seems not be well tested.

the ra8876 driver can still be used but must be used together with USE_UNIVERSAL_TOUCH.
the universal touch driver is now integrated into the RA8876 driver

**Reply by helmar74** [reply]
OK will give a try. I'm not using touch at the moment. Do I still have to use USE_UNIVERSAL_TOUCH

**Reply by gemu2015** [reply]
no, if not using touch you don't need it

btw this is the menuconfig flag responsible for the jpeg crash:

CONFIG_ESP_ROM_HAS_JPEG_DECODE

it has not been defined in core2

**Comment by Jason2866** [comment]
`CONFIG_ESP_ROM_HAS_JPEG_DECODE` is not an option which is configurable via menuconfig. It is set "hard" by IDF

**Comment by helmar74** [comment]
I'll have a look on the RA8876 Display issue. As far as I can see, the device will not be updated anymore to uDisplay, as it is deeply integrated in my openhab system an a lot of pictures are display. Maybe I start digging into universal display and I create a display.ini for it.

One other thing:
I downloaded current development version and unpacked it into a directory. As I'm also runnin ESP32C3 I only changed platform_overiride.ini in that way:
```
;                          tasmota
;                          tasmota-debug
;                          tasmota-minimal
;                          tasmota-lite
;                          tasmota-knx
;                          tasmota-sensors
;                          tasmota-display
;                          tasmota-zbbridge
;                          tasmota-ir
;                          tasmota32
;                          tasmota32solo1
;                          tasmota32s2
;                          tasmota32s2cdc
;                          tasmota32s3
;                          tasmota32c2
                          tasmota32c3
;                          tasmota32c6
```

user_override_config.h is completly unchanged.
I get the following compile error:
Compiling .pio\build\tasmota32c3\lib050\ESP8266Audio\AudioFileSourceBuffer.cpp.o
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/users/helmar/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/12.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1

Do I have to install something further to compile, or why is the error occuring?

or is there an error in tasmota source/platformio.
The directory looks like this:
<img width="599" alt="image" src="https://github.com/arendst/Tasmota/assets/25499089/e7ad02b2-f07e-4c86-a17f-a400b9df958e">

**Reply by Jason2866** [reply]
There is no error in Tasmota Platformio Setup. I dont know why you are digging in `platformio/packages` There is absolute nothing which should be changed from user.
If such strange error arrives the Platformio Setup has been screwed by user. The default config setup works.

**Reply by helmar74** [reply]
I didn'T changed anything.
Strange but why is this error occuring. I deinstalled platformio addon in vsc, delete .platformio folder an newly unzipped current development. 
than the error comes up...

**Reply by Jason2866** [reply]
idk what is broken in your setup

**Reply by helmar74** [reply]
I tried it on a new windows machine. There visual studio code nor platformio was installed. 
The same error comes up:
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/users/helmar/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/12.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
Archiving .pio\build\tasmota32c3\lib4c7\libWiFiHelper.a
*** [.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
======================================================================================================================================== [FAILED] Took 269.27 seconds ========================================================================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32c3    FAILED    00:04:29.274
=================================================================================================================================== 1 failed, 0 succeeded in 00:04:29.274 =================================================================================================================================== 

 *  The terminal process "C:\Users\helmar\.platformio\penv\Scripts\platformio.exe 'run'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it.

**Reply by helmar74** [reply]
@Jason2866 : Are you able co compile current development fot tasmota32c3 on a Windows PC with VSC and platformio add on?

**Comment by Jason2866** [comment]
@helmar74 I don't have a Windows PC. C3 compiles without problems on my Mac and it compiles in Github Actions CI with Linux. Every commit is compiled and tested. So the C3 too. One build variant is compiled in CI with Windows.
Theo (the project owner) is using Windows. If there is an issue he will tell me ;-)

**Reply by helmar74** [reply]
ok, maybe it is only a problem compiling on windows

**Reply by arendst** [reply]
Oh @helmar74 you're inresistable. It compiles fine on my windows.

**Reply by helmar74** [reply]
ok, not sure what is happening on my 2 PC's?
Did you compile tasmota32c3 ?

**Reply by arendst** [reply]
YOU'RE RIGHT! It fails to compile. Let me chk.....

**Reply by arendst** [reply, marked as answer]
OK. Update your Tasmota package with latest dev release.

File ``platformio_tasmota32.ini`` was updated to a newer platform. This fixes compilation on Windows too (at least it fixed it for me).

**Reply by helmar74** [reply]
I tried it it version https://github.com/arendst/Tasmota/pull/21381[)](https://github.com/arendst/Tasmota/commit/da0e4c433f607bddeb6753ead36c470f9939da12)

Now it compiled ;-)

<img width="207" alt="image" src="https://github.com/arendst/Tasmota/assets/25499089/ccf77a3f-9ba8-471e-b143-d11817213de3">

**Comment by helmar74** [comment]
I'll also give it a try with my display version tomorrow. I keep you updated here.

**Comment by helmar74** [comment]
Hi, tasmota32c3 is now compiling for me. @arendst : Thanks for your assistance:

Now back to my initial issue. 
I'm running a RA8876 Display (without touch function) on a ESP32.
I created a user_config_override for this. I also need JPGs on the Display,
so the following define is in user_config_override.h, to make display of JPGs possible
```
#define JPEG_PICTS 
```

With this define it was working well until Tasmota stable 13.4.0.
Now I get a compile error, when compiling tasmota32:
```
Compiling .pio\build\tasmota32\src\tasmota.ino.cpp.o
In file included from tasmota/my_user_config.h:1302,
                 from D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota.ino:27:
tasmota/user_config_override.h:24:2: warning: #warning **** user_config_override.h: Using Settings from this File **** [-Wcpp]
   24 | #warning **** user_config_override.h: Using Settings from this File ****
      |  ^~~~~~~
D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_support/support_jpeg.ino: In function 'bool jpg2rgb888(const uint8_t*, size_t, uint8_t*, jpg_scale_t)':
D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_support/support_jpeg.ino:135:39: error: invalid conversion from 'uint32_t (*)(void*, size_t, uint8_t*, size_t)' {aka 'long unsigned int (*)(void*, unsigned int, unsigned char*, unsigned int)'} to 'jpg_reader_cb' {aka 'unsigned int (*)(void*, unsigned int, unsigned char*, unsigned int)'} [-fpermissive]
  135 |     if(esp_jpg_decode(src_len, scale, _jpg_read, _rgb_write, (void*)&jpeg) != ESP_OK){
      |                                       ^~~~~~~~~
      |                                       |
      |                                       uint32_t (*)(void*, size_t, uint8_t*, size_t) {aka long unsigned int (*)(void*, unsigned int, unsigned char*, unsigned int)}
In file included from C:/Users/helmar/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32/include/esp32-camera/conversions/include/img_converters.h:25,
                 from D:/Users/Helmut/Desktop/1/Tasmota/Tasmota-development_v13412/tasmota/tasmota_support/support_jpeg.ino:24:
C:/Users/helmar/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32/include/esp32-camera/conversions/include/esp_jpg_decode.h:37:71: note:   initializing argument 3 of 'esp_err_t esp_jpg_decode(size_t, jpg_scale_t, jpg_reader_cb, jpg_writer_cb, void*)'        
   37 | esp_err_t esp_jpg_decode(size_t len, jpg_scale_t scale, jpg_reader_cb reader, jpg_writer_cb writer, void * arg);
      |                                                         ~~~~~~~~~~~~~~^~~~~~
*** [.pio\build\tasmota32\src\tasmota.ino.cpp.o] Error 1
Compiling .pio\build\tasmota32\lib2ae\WiFiHelper\WiFiHelper_ESP32.cpp.o
In file included from include/../tasmota/my_user_config.h:1302,
                 from include/tasmota_options.h:23,
                 from lib/default/WiFiHelper/src/WiFiHelper_ESP32.cpp:34:
include/../tasmota/user_config_override.h:24:2: warning: #warning **** user_config_override.h: Using Settings from this File **** [-Wcpp]
   24 | #warning **** user_config_override.h: Using Settings from this File ****
```

Does anyone have an idea, what I can try, to get JPGs working on current development 13.4.1.2?

**Reply by gemu2015** [reply]
as already said, core3 breaks the jpeg decoder. it will crash because the ROM routine does not work. you will have to wait until this is fixed in core3, probably by removing CONFIG_ESP_ROM_HAS_JPEG_DECODE from sdkconfig

however core3 compile will work with this small fix 
[support_jpeg.ino.zip](https://github.com/arendst/Tasmota/files/15247379/support_jpeg.ino.zip)

**Comment by helmar74** [comment]
Thank you, I'll try it. But will this fix also be in Tasmota 14.0.0?

**Comment by Jason2866** [comment]
`CONFIG_ESP_ROM_HAS_JPEG_DECODE` is not an option which can be configured in IDF menuconfig.
It is there, as long espressif does not change this setting. There is no change expected anytime soon.

**Comment by helmar74** [comment]
I can confirm, that I was able to compile and now JPEGS are also working.
So would be great to have the change it next stable version of tasmota

**Comment by helmar74** [comment]
I can confirm, that after @arendst integrated the fix into current development now works

[https://github.com/arendst/Tasmota/commit/61814e79feca9eb8d31baf577ee3176a30cce424](https://github.com/arendst/Tasmota/commit/61814e79feca9eb8d31baf577ee3176a30cce424)

Thanks a lot

helmar
